### PR TITLE
feature/mx-1569 date validation

### PR DESCRIPTION
**Pull request by carolinscholl**
# PR Context
In json schema, datetime accepts only dates with a timestamp (yyyy-MM-dd HH:mm:ss), and date accepts only dates in the format "yyyy-MM-dd". However, we also want to validate partial dates (e.g. yyyy-MM), in order to fulfill the type requirements specified in mex-model. That is why we subclass the Timestamp class into "dates" up to different resolutions. 

# Changes
- Renamed Timestamp class to TemporalEntity and created three subclasses with specific resolution YearMonth, YearMonthDay, YearMonthDayTime

# Fixed
- temporal entity validation is in place again and harmonized with mex-model

### CODE REVIEW

File: mex/common/types/timestamp.py
Comment (cutoffthetop) [inline mex/common/types/timestamp.py:324]:
```suggestion
        if self.precision not in TIME_PRECISIONS:
            raise ValueError("Expected time-based precision level")
```

i think we could/should allow the hour/minute/ms here too

Comment (carolinscholl) [reply mex/common/types/timestamp.py:324]: hmm but like this we allow all precisions, even year etc.?

Comment (cutoffthetop) [reply mex/common/types/timestamp.py:324]: should just be these 4, or am i missing sth?

```python
TIME_PRECISIONS = [
    TimestampPrecision.HOUR,
    TimestampPrecision.MINUTE,
    TimestampPrecision.SECOND,
    TimestampPrecision.MICROSECOND,
]
```

Comment (carolinscholl) [reply mex/common/types/timestamp.py:372]: you're absolutely right, thought it includes all precisions, will change it!

File: mex/common/types/temporal_entity.py
Comment (cutoffthetop) [inline mex/common/types/temporal_entity.py:323]: "accepts" might be a bit misleading, you can still parse from any format that describes a day-precision _temporal entity_, e.g. `YearMonthDay("1st June 2020")`

File: mex/common/types/temporal_entity.py
Comment (cutoffthetop) [inline mex/common/types/temporal_entity.py:362]:
```suggestion
                core_schema.is_instance_schema(cls),
```

File: mex/common/types/temporal_entity.py
Comment (cutoffthetop) [inline mex/common/types/temporal_entity.py:352]: maybe put this as a class var `STR_SCHEMA_PATTERN = YEAR_MONTH_DAY_REGEX` and then we can have just 1 `__get_pydantic_core_schema__` instead of 4

File: mex/common/types/temporal_entity.py
Comment (cutoffthetop) [inline mex/common/types/temporal_entity.py:332]: we could just put `ALLOWED_PRECISION_LEVELS` as a class var and then we could move this check to `TemporalEntity` and wouldn't have to have an `__init__` on the subclasses at all

File: tests/models/test_model_schemas.py
Comment (cutoffthetop) [inline tests/models/test_model_schemas.py:132]: nice 🙏

File: mex/common/models/resource.py
Comment (cutoffthetop) [inline mex/common/models/resource.py:62]: i'm liking the self-explanatory look of this :)

**Review by cutoffthetop — CHANGES_REQUESTED** [review]
really nice!

just some code duplication we could axe in the TemporalEntity subclasses. -- and then we have to wait for the mex-model approval of course

File: mex/common/types/temporal_entity.py
Comment (carolinscholl) [reply mex/common/types/temporal_entity.py]: I changed the docstrings: https://github.com/robert-koch-institut/mex-common/commit/9000245148d2f6e4566b573d515beba459fb27b5

File: mex/common/types/temporal_entity.py
Comment (carolinscholl) [reply mex/common/types/temporal_entity.py]: Done: https://github.com/robert-koch-institut/mex-common/commit/57c0b499a69f46a0a0a7fdc75a7a5b5c8fd94c63

File: mex/common/types/temporal_entity.py
Comment (carolinscholl) [reply mex/common/types/temporal_entity.py]: Great idea, done in: https://github.com/robert-koch-institut/mex-common/commit/57c0b499a69f46a0a0a7fdc75a7a5b5c8fd94c63

File: mex/common/types/temporal_entity.py
Comment (carolinscholl) [reply mex/common/types/temporal_entity.py]: Another great idea, done in https://github.com/robert-koch-institut/mex-common/commit/1f6574f6ca9d6d9ae00852e44f51d79634ed3f5b

**Review by cutoffthetop — DISMISSED** [review]

**Pull request merged**
